cryptography/src/bls12381/primitives/ops.rs

```diff
@@ -216,6 +216,81 @@ where
     V::verify_prehashed(&public, &hm_sum, &signature)
 }
 
+/// Verifies multiple [PartialSignature]s over the same message as a single
+/// aggregate signature (or returns any invalid signature found).
+///
+/// # Warning
+///
+/// This function assumes a group check was already performed on each `signature`.
+pub fn partial_verify_multiple_public_keys<'a, V, I>(
+    public: &poly::Public<V>,
+    namespace: Option<&[u8]>,
+    message: &[u8],
+    partials: I,
+) -> Result<(), Vec<PartialSignature<V>>>
+where
+    V: Variant,
+    I: IntoIterator<Item = &'a PartialSignature<V>>,
+{
+    // Collect public keys for partial signatures
+    let mut public_keys = Vec::new();
+    let mut signatures = Vec::new();
+    let mut collected = Vec::new();
+    for partial in partials {
+        public_keys.push(public.evaluate(partial.index).value);
+        signatures.push(partial.value);
+        collected.push(partial);
+    }
+
+    // Compute aggregate public key
+    let aggregate_public_key = aggregate_public_keys::<V, _>(&public_keys);
+
+    // Compute aggregate signature
+    let aggregate_signature = aggregate_signatures::<V, _>(&signatures);
+
+    // Verify message over computed aggregate representations
+    if verify_message::<V>(
+        &aggregate_public_key,
+        namespace,
+        message,
+        &aggregate_signature,
+    )
+    .is_ok()
+    {
+        return Ok(());
+    }
+
+    // If verify fails, find offending signatures via recursive bisection
+    if collected.len() <= 1 {
+        return Err(collected.into_iter().cloned().collect());
+    }
+    let mid = collected.len() / 2;
+    let (left, right) = collected.split_at(mid);
+    let result_l = partial_verify_multiple_public_keys::<V, _>(
+        public,
+        namespace,
+        message,
+        left.iter().copied(),
+    );
+    let result_r = partial_verify_multiple_public_keys::<V, _>(
+        public,
+        namespace,
+        message,
+        right.iter().copied(),
+    );
+
```

Comment:
Considered parallelizing this but figure let's just focus on the simple case for now.